### PR TITLE
Legend: Render legend threshold colors

### LINKED ISSUE #92184: Legend: Refactor legend component for Bar Chart value mapping and threshold colors
The current legend component needs to be improved to handle multiple color sources effectively, starting from Bar chart legend. This task involves refactoring the legend to ensure it can dynamically and accurately represent value mappings and threshold colors associated with different data sources.

Additional Notes:

- Here is the dashboard to address the problem: [BarChart - Thresholds & Mappings](https://play.grafana.org/d/barchart-thresholds-mappings/barchart-thresholds-and-mappings).
- Collaboration with the UX/UI team is necessary to maintain design consistency across different chart types. Here is [figma doc](https://www.figma.com/design/inbyYWBO57D7DNLe7kNiZC/Legends?node-id=9-40&t=AUitIs4nqF9JTlqJ-0) with ui/ux grooming.

### PR DESCRIPTION
**Pull request by Develer**
**What is this feature?**
The current legend component needs to be improved to handle multiple color sources effectively, starting from Bar chart legend. This task involves refactoring the legend to ensure it can dynamically and accurately represent value mappings and threshold colors associated with different data sources.

Please check `dev-dashboards -> BarChart - Thresholds & Mappings` dashboard to see improved value mappings and thresholds colors.

**Which issue(s) does this PR fix?**:

Fixes #92184 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/VizLegend/VizLegend.tsx [hunk 1]
```diff
@@ -104,21 +104,40 @@ export function VizLegend<T>({
         />
       );
     case LegendDisplayMode.List:
+      const isThresholdsEnabled = thresholdItems && thresholdItems.length > 1;
+      const isValueMappingEnabled = mappingItems && mappingItems.length > 0;
       return (
         <>
+          {/* render items when single series and there is no thresholds and no value mappings
+           * render items when multi series and there is no thresholds
```
Comment (baldm0mma) [inline packages/grafana-ui/src/components/VizLegend/VizLegend.tsx:112]: Thanks for the comment here!

File: public/app/plugins/panel/barchart/BarChartLegend.tsx [hunk 2]
```diff
@@ -32,17 +31,37 @@ export const BarChartLegend = memo(
...
-      const items = getFieldLegendItem([colorField], theme);
+    const fieldConfig = data[0].fields[0].config;
+    const colorMode = fieldConfig.color?.mode;
+
+    const thresholdItems: VizLegendItem[] = [];
+    if (colorMode === FieldColorModeId.Thresholds) {
+      for (let i = 0; i < data[0].fields.length; i++) {
+        const thresholds = data[0].fields[i].config.thresholds;
+        if (thresholds) {
+          // in case of multiple fields, each field has same thresholds config as in first field (data[0].field[0].config),
```
Comment (baldm0mma) [inline public/app/plugins/panel/barchart/BarChartLegend.tsx:42]: 👍

**Review by baldm0mma — APPROVED** [review]

**Review by codeincarnate — COMMENTED** [review]
Looking good! 

I'm assuming this should support gradients as well?

Screenshot 2024-09-25 at 1 39 12 PM

**Comment by adela-almasan** [comment]
@Develer good job on this one, it looks really put together! 🎉 

What I've noticed is that we're loosing the ability to override colors from legend. And not sure if it ever worked, but the usual functionality to click and hide/show doesn't seem to work either.

https://github.com/user-attachments/assets/177b866d-63e8-4ba3-8afb-4fafae3c263c

**Comment by Develer** [comment]
> Looking good!
> 
> I'm assuming this should support gradients as well?
> 
> Screenshot 2024-09-25 at 1 39 12 PM

The gradient should left without changing. Thresholds and value mappings colors the only parts that updated.

**Comment by Develer** [comment]
> @Develer good job on this one, it looks really put together! 🎉
> 
> What I've noticed is that we're loosing the ability to override colors from legend. And not sure if it ever worked, but the usual functionality to click and hide/show doesn't seem to work either.
> 
>  legend_colors.mov

About override colors from legend - we still can override from legend but series color only, we can't override thresholds or value mappings from legends (this is true for `main` also).

For show/hide click functionality - it's working, it's just a single series examples, that's why you don't see any effect.

**Review by adela-almasan — APPROVED** [review]
:shipit:

**Pull request merged**